Blog > Application: Create 포트, 서비스 및 여러 Use Case 추가

merge-simpson opened the pull request:
## Issues

- Resolves #42 
- Resolves #53 
<!-- 이 PR이 완전히 처리한 이슈의 번호를 목록으로 작성합니다. PR 병합 시 목록의 이슈들이 자동으로 close 됩니다. -->
<!-- 다른 레포지터리의 이슈: `Resolves nettee-space/another-repository#0` -->

<br />

## Description

- Task of #9 &nbsp; **>** &nbsp; #43 
- Command port에 `transaction(() -> {})` 함수 지원
- Use-case는 미리 개발하기로 논의된 목록을 작성했습니다.
- Create 관련 port 및 service를 작성했습니다.

<br />

## Review Points

<!-- 리뷰어가 중점적으로 확인할 항목을 작성해 주세요. -->
<!-- (Ex: 구현 로직, API 스펙, 테스트 케이스 등) -->

- `create` 및 `save` 위주로 검토해 주시면 좋을 것 같습니다.
- `transaction` 함수는 이하 방식으로 설계를 시도해 보려고 합니다. (아직 사용하지 않음.)  
    ```java
    port.transaction(() -> {
        port.함수1();
        port.함수2();
        port.함수3();
    });
    ```
- 테스트 케이스  
  - ✅ 정상 등록
  - 🚧 예외 반환: 사용자당 블로그 개설 상한(1개) 초과 시 예외 반환

<br />

## How Has This Been Tested?

- Kotest `FreeSpec` 테스트코드 통과 확인

<br />

merge-simpson committed:
feat(blog): add basic crud use-cases about blog

merge-simpson committed:
feat(blog): add subscription use-cases about blog, newsletter

merge-simpson committed:
feat(blog): add save in command port

merge-simpson committed:
feat(blog): add transaction(()->{}) in command port

merge-simpson committed:
feat(blog): implement save in command service

merge-simpson committed:
test(blog): add normal 'create' test

merge-simpson committed:
chore(blog): fix and add comments

merge-simpson commented on services/blog/application/src/main/java/nettee/blolet/blog/application/port/BlogCommandPort.java:
```diff
@@ -0,0 +1,9 @@
+
+import nettee.blolet.blog.domain.Blog;
+
+public interface BlogCommandPort {
+    void transaction(Runnable runnable);
```
리뷰하실 때 예상 용례를 참고해 주세요:

```java
port.transaction(() -> {
    port.함수1();
    port.함수2();
    port.함수3();
});
```

wch-os replied:
와.. 안 그래도 service 계층에서 @Transactional을 사용해야 되나 고민하고 있었는데, 이런 방법이 있었네요..!!
runnable을 이렇게 사용해서 service에서 spring 의존성 하나 없이 순수하게 유지할 수 있네요👍

silberbullet replied:
귀한 정보 감사드립니다! 저도 적용해서 사용할 것 같습니다!

wowddok99 reviewed (approved):
올려주신 코드 확인하였습니다!
고생하셨습니다😄

merge-simpson committed:
fix(blog): override equals, hashCode in domain

merge-simpson committed:
feat(blog): add `BLOG_MAXIMUM_EXCEEDED` error code

merge-simpson committed:
feat(blog): add `countByUserId` in command port

merge-simpson committed:
feat(blog): throw exception in service if blog count per user maximum exceeded (creation)

merge-simpson committed:
refactor(blog): rename test class excluding underscored suffix

merge-simpson committed:
test(blog): add `userId` to mocking object

merge-simpson committed:
test(blog): refactor normal creation test code

merge-simpson committed:
test(blog): add test to throw blog maximum exceeded error when user exceeds blog creation limit

use kotest shouldThrow to assert BLOG_MAXIMUM_EXCEEDED when countByUserId meets limit and mock save behavior

merge-simpson commented on services/blog/application/src/test/kotlin/nettee/blolet/blog/application/service/BlogCommandServiceTest.kt:
```diff
@@ -0,0 +1,97 @@
+            verify(exactly = 1) { commandPort.save(item) }
+        }
+
+        "🚧 사용자당 블로그 개설 개수를 초과하면 오류를 반환한다." {
+            // mock:
+            val capturedItem = mutableListOf<Blog>(item)
+            every { commandPort.countByUserId(any()) } answers {
+                val userIdArg: String = firstArg()
+                capturedItem.count {it.userId == userIdArg}
+            }
+            every { commandPort.save(capture(capturedItem)) } answers {
+                val inputtedBlog: Blog = firstArg()
+                Blog.builder()
+                    .id(inputtedBlog.id ?: "1")
+                    .userId(inputtedBlog.userId)
+                    .name(inputtedBlog.name)
+                    .url(inputtedBlog.url)
+                    .createdAt(inputtedBlog.createdAt)
+                    .updatedAt(inputtedBlog.createdAt)
+                    .build()
+            }
+
+            // action & assert:
+            val exception = shouldThrow<CustomException> {
+                commandService.save(item)
+            }
+            exception.errorCode shouldBe BLOG_MAXIMUM_EXCEEDED
+        }
+    }
```
💬 테스트 설명

> 준비: `capturedItems` 초깃값으로, 파라미터에 `item`을 미리 넣었습니다. (`mutableListOf<Blog>(item)`)

- `capturedItems`에 이미 사용자의 블로그 `item`이 존재합니다.
- 이후 `save` 시도 시, `port.countByUserId(...)` 반환 결과가 이미 1입니다.
- **서비스 로직에서 이때 예외를 잘 응답하는지 확인합니다.**

merge-simpson commented:
@wowddok99 주요 부분에서 수정이 있어 리뷰를 다시 요청했습니다!

merge-simpson committed:
refactor(blog): rename `-Port` to `-RepositoryPort`

wowddok99 reviewed (approved):
추가로 올려주신 코드도 확인하였습니다!

capturedItems에 미리 item을 넣어 사용자 블로그가 이미 1개 존재하는 상황을 미리 설명해주셔서 리뷰하기 수월했습니다! 
고생하셨습니다😄

wch-os reviewed (approved):
깔끔하고 표본이 될 만한 PR인 것 같습니다!!👏

수고하셨습니다!

silberbullet commented on services/blog/application/src/main/java/nettee/blolet/blog/application/port/BlogCommandPort.java:
```diff
@@ -0,0 +1,11 @@
+
+import nettee.blolet.blog.domain.Blog;
+
+public interface BlogCommandPort {
```
###  💊  가이드 맞게 Port 인터페이스 수정 부탁드립니다.

ex)
- 직접적 데이터 조작과 연관 PORT: `-RepositoryPort`
- Client와 연관 PORT: `-NetteeClientPort`

merge-simpson replied:
말씀해 주신 대로 하기 규칙을 적용했습니다.
예리한 관찰 감사합니다. 👍 

- DB 보존과 관련한 조작 port: `-RepositoryPort`
- S2S client 역할 port: `-NetteeClientPort`

silberbullet reviewed (changes requested):
### 빠른 작업 감사드립니다! 👍 

테스트 코드의 테스트 케이스명에 이모지를 넣은 점이 인상적이었습니다.
다만, 처음 리뷰자가 자각과 기대하는 의도와 다르게 받아들여질 수도 있겠다는 생각이 들었습니다.

초반 테스트 코드는 각자의 방식대로 작성해보셔도 좋을 것 같습니다!

Interface 명만 변경된다면 Merge해도 괜찮을 것 같습니다.

silberbullet reviewed (approved):
빠른 작업 감사드립니다! Merge 해도 좋습니디!
